### PR TITLE
[libc++] Clean up includes of <__assert>

### PR DESCRIPTION
**Pull request by ldionne**
Originally, we used __libcpp_verbose_abort to handle assertion failures. That function was declared from all public headers. Since we don't use that mechanism anymore, we don't need to declare __libcpp_verbose_abort from all public headers, and we can clean up a lot of unnecessary includes.

This patch also moves the definition of the various assertion categories to the <__assert> header, since we now rely on regular IWYU for these assertion macros.

rdar://105510916

### CODE REVIEW

File: libcxx/test/libcxx/assertions/modes/none.pass.cpp [hunk 1]
```diff
@@ -6,11 +6,12 @@
 //
 //===----------------------------------------------------------------------===//
 
-// This test checks that if no hardening mode is defined (i.e., in the unchecked mode), by default assertions aren't
-// triggered.
+// This test checks that if no hardening mode is defined (i.e., in the unchecked mode),
```
Comment (var-const) [inline libcxx/test/libcxx/assertions/modes/none.pass.cpp:9]: Nit: this reformatting seems unrelated.

File: libcxx/include/vector [hunk 2]
```diff
@@ -831,8 +831,8 @@ private:
   // For more details, see the "Using libc++" documentation page or
   // the documentation for __sanitizer_annotate_contiguous_container.
 
-  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __annotate_contiguous_container(
-      const void* __old_mid, const void* __new_mid) const {
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void
```
Comment (var-const) [inline libcxx/include/vector:834]: Nit: this reformatting seems unrelated.

File: libcxx/include/valarray [hunk 3]
```diff
@@ -2435,7 +2434,7 @@ template <class _Expr, __enable_if_t<__is_val_expr<_Expr>::value, int> >
 inline valarray<_Tp>& valarray<_Tp>::operator*=(const _Expr& __v) {
   size_t __i = 0;
   for (value_type* __t = __begin_; __t != __end_; ++__t, ++__i)
-    *__t *= std::__get(__v,__i);
+    *__t *= std::__get(__v, __i);
```
Comment (var-const) [inline libcxx/include/valarray:2437]: Nit: this reformatting seems unrelated.

File: libcxx/include/string [hunk 4]
```diff
@@ -938,7 +938,11 @@ public:
       // Turning off ASan instrumentation for variable initialization with _LIBCPP_STRING_INTERNAL_MEMORY_ACCESS
       // does not work consistently during initialization of __r_, so we instead unpoison __str's memory manually first.
       // __str's memory needs to be unpoisoned only in the case where it's a short string.
-      : __r_([](basic_string &__s) -> decltype(__s.__r_)&& { if(!__s.__is_long()) __s.__annotate_delete(); return std::move(__s.__r_); }(__str)) {
+      : __r_([](basic_string& __s) -> decltype(__s.__r_)&& {
```
Comment (var-const) [inline libcxx/include/string:941]: Nit: this reformatting seems unrelated.

File: libcxx/include/scoped_allocator [hunk 5]
```diff
@@ -476,8 +475,8 @@ public:
   }
 
 private:
-  _LIBCPP_HIDE_FROM_ABI explicit scoped_allocator_adaptor(outer_allocator_type&& __o, inner_allocator_type&& __i) _NOEXCEPT
-      : base(std::move(__o), std::move(__i)) {}
+  _LIBCPP_HIDE_FROM_ABI explicit scoped_allocator_adaptor(
```
Comment (var-const) [inline libcxx/include/scoped_allocator:478]: Nit: this reformatting seems unrelated.

**Review by var-const — APPROVED** [review]
LGTM (just a couple nits re. some potentially unrelated reformatting).

**Comment by MaskRay** [comment]
`libcxx/src/new.cpp` needs `#include <__assert>` in the `_LIBCPP_ASSERT_SHIM` code path ( `-fno-exceptions` and LIBCXX_ENABLE_NEW_DELETE_DEFINITIONS=on). Sent  #83480 as a follow-up.
